src/qminer/qminer_core.cpp

```diff
@@ -2092,49 +2237,66 @@ PRecSet TRec::ToRecSet() const {
-	// get join info
-	AssertR(Store->IsJoinId(JoinId), "Wrong Join ID");
-	const TJoinDesc& JoinDesc = Store->GetJoinDesc(JoinId);
-	// check if index join
-	if (JoinDesc.IsIndexJoin()) {
-		TUInt64IntKdV JoinRecIdFqV;
-		if (IsByRef()) {
-			// by reference
-			Assert(Store->IsRecId(GetRecId()));
-			// do join using index
-			const int JoinKeyId = JoinDesc.GetJoinKeyId();
-			// prepare join query
-			TIntUInt64PrV JoinQueryV = TIntUInt64PrV::GetV(TIntUInt64Pr(JoinKeyId, GetRecId()));
-			// execute join query
-			Base->GetIndex()->SearchOr(JoinQueryV, JoinRecIdFqV);
-		} else {
-			// do join using serialized record set
-			if (JoinIdPosH.IsKey(JoinId)) {
-				const int Pos = JoinIdPosH.GetKey(JoinId);
-				TMIn MIn(RecVal.GetBf() + Pos, RecVal.Len() - Pos, false);
-				JoinRecIdFqV.Load(MIn);
-			}
-		}
-		return TRecSet::New(JoinDesc.GetJoinStore(Base), JoinRecIdFqV);
-	} else if (JoinDesc.IsFieldJoin()) {
-		// do join using store field
-		const int JoinRecFieldId = JoinDesc.GetJoinRecFieldId();
-		const uint64 JoinRecId = IsFieldNull(JoinRecFieldId) ? TUInt64::Mx : GetFieldUInt64(JoinRecFieldId);
-		// get join weight
-		const int JoinFqFieldId = JoinDesc.GetJoinFqFieldId();
-		const int JoinRecFq = IsFieldNull(JoinRecFieldId) ? 0 : GetFieldInt(JoinFqFieldId);
-		// return record set
-		if (JoinRecId != TUInt64::Mx) {
-			// return record
-			return TRecSet::New(JoinDesc.GetJoinStore(Base), JoinRecId, JoinRecFq);
-		} else {
-			// no record, return empty set
-			return TRecSet::New(JoinDesc.GetJoinStore(Base));
-		}
-	}
-	// unknown join type
-	throw TQmExcept::New("Unsupported join type for join " + JoinDesc.GetJoinNm() + "!");
+    // get join info
+    QmAssertR(Store->IsJoinId(JoinId), "Wrong Join ID");
+    const TJoinDesc& JoinDesc = Store->GetJoinDesc(JoinId);
+    // check if index join
+    if (JoinDesc.IsIndexJoin()) {
+        TUInt64IntKdV JoinRecIdFqV;
+        if (IsByRef()) {
+            // by reference
+            Assert(Store->IsRecId(GetRecId()));
+            // do join using index
+            const int JoinKeyId = JoinDesc.GetJoinKeyId();
+            // prepare join query
+            TIntUInt64PrV JoinQueryV = TIntUInt64PrV::GetV(TIntUInt64Pr(JoinKeyId, GetRecId()));
+            // execute join query
+            Base->GetIndex()->SearchOr(JoinQueryV, JoinRecIdFqV);
+        } else {
+            // do join using serialized record set
+            if (JoinIdPosH.IsKey(JoinId)) {
+                const int Pos = JoinIdPosH.GetKey(JoinId);
+                TMIn MIn(RecVal.GetBf() + Pos, RecVal.Len() - Pos, false);
+                JoinRecIdFqV.Load(MIn);
+            }
+        }
+        return TRecSet::New(JoinDesc.GetJoinStore(Base), JoinRecIdFqV);
+    } else if (JoinDesc.IsFieldJoin()) {
+        // do join using store field
+        const int JoinRecFieldId = JoinDesc.GetJoinRecFieldId();
+        const uint64 JoinRecId = IsFieldNull(JoinRecFieldId) ? TUInt64::Mx : GetFieldUInt64Safe(JoinRecFieldId);
```

Comment:
Add function GetFieldJoinRecId to TStore that combines above two lines. Join provided as Id, Name, Desc.